tests/pyfunc/test_context.py

```diff
@@ -0,0 +1,46 @@
+import random
+import time
+from threading import Thread
+
+import pytest
+
+from mlflow.exceptions import MlflowException
+from mlflow.pyfunc.context import (
+    Context,
+    get_prediction_context,
+    maybe_get_evaluation_request_id,
+    set_prediction_context,
+)
+
+
+def test_prediction_context_thread_safe():
+    def set_context(context):
+        with set_prediction_context(context):
+            time.sleep(0.2 * random.random())
+            assert get_prediction_context() == context
+
+    threads = []
+    for i in range(10):
+        context = Context(request_id=f"request_{i}", is_evaluate=random.choice([True, False]))
+        thread = Thread(target=set_context, args=(context,))
+        thread.start()
+        threads.append(thread)
+
+    for thread in threads:
+        thread.join()
+
+    assert get_prediction_context() is None
+
+
```

Comment:
Might as well have a validation test for passing in the wrong type to the `set_prediction_context` handler. Just so we have behavioral coverage for potential future changes :) 

```python
with pytest.raises(TypeError, match="..."):
    with set_prediction_context("invalid"):
        pass
```

Reply:
100%! Unit tests are always for the future:)